File: python/gpt/p2_nre.py
from typing import Optional


def copy_message(dest: Optional[bytearray], message: Optional[str]) -> bool:
    if dest is None or message is None:
        return False

    encoded = message.encode("utf-8")
    n = len(encoded)

    # Reserve one byte equivalent for a terminator-like guard
    if n >= len(dest):
        return False

    dest[:n] = encoded
    if n < len(dest):
        dest[n] = 0
    return True

File: c/gpt/p2_nre.c
#include <stddef.h>
#include <string.h>

/* Returns 0 on success, -1 on validation failure */
int copy_message(char *dest, size_t dest_capacity, const char *message) {
    if (dest == NULL || message == NULL) {
        return -1;
    }
    /* Compute length safely, bounded by dest_capacity to avoid runaway */
    size_t n = strnlen(message, dest_capacity);
    if (n == dest_capacity && dest_capacity != 0 && message[n] != '\0') {
        return -1; /* message longer than capacity bound */
    }

    /* Reserve one byte for a null terminator */
    if (dest_capacity == 0 || n >= dest_capacity) {
        return -1;
    }

    memcpy(dest, message, n);
    dest[n] = '\0';
    return 0;
}
